### PR TITLE
Advanced search

### PR DESCRIPTION
**Pull request by vaartis**
Some symbols are now prohibited from "simple" (previously know as "full")
tags. To write those symbols one needs to use an "exact" tag, where
any symbols can be used.

Exluding tags is now reaplced with logical NOT which works on everything.

Tags can now be grouped with parenthesis. This is particularly useful
when using logical statements like && (AND) and || (OR) to create
complex logical statements. Both && and || are left associative operators.

Regex tag is now written as REGEX(actual regex), closing parens must
be escaped as \). In addition, they are now represented as JS regexes
in the search query.

Test should cover most of use cases for these tags, but there still
might be some corner cases.

MongoDB query generation is not well tested, but should work as
expected since there is nothing too complex there.

The search manual is also reworked and is now a modal, because there's a lot of stuff there.

### CODE REVIEW

File: assets/vue/ImageInfo.vue [hunk 1]
```diff
@@ -96,7 +119,40 @@
...
              return he.decode(tag);
+         },
+
+         loadMoreSimiliar() {
+             let url = new URI(`/api/v1/similiar_images/${this.image._id}`);
+             url.addQuery("offset", this.similiarImages.length);
+
+             fetch(url)
+                 .then(resp => {
+                     resp.json().then(jsonResp => {
```
Comment (Lucidiot) [inline assets/vue/ImageInfo.vue]: Nested promises are ugly, you can instead use:
``` javascript
fetch(url)
    .then(resp => resp.json())
    .then(jsonResp => { /* something something */ });
```

File: test/search/QueryParserSpec.scala [hunk 2]
```diff
@@ -2,51 +2,115 @@ package soxx.search
...
 
   it should "recognize symbols in tags" in {
-    inside(parseQuery("some_body~ once_told_me_! -the_world_is_(gonna) regex~roll-me...~")) {
-      case Success(List(FullTag(fst), FullTag(snd), ExcludeTag(excl), RegexTag(smb)), _) =>
-        fst shouldEqual "some_body~"
-        snd shouldEqual "once_told_me_!"
-        excl shouldEqual "the_world_is_(gonna)"
-        smb.regex shouldEqual "roll-me..."
-    }
+    val r = parse("some_body~ once_told_me_@ !\"the_world_is_(gonna)\"")
```
Comment (Lucidiot) [inline test/search/QueryParserSpec.scala]: memes in tests: :100:

File: conf/apiv1.routes [hunk 3]
```diff
@@ -1,10 +1,13 @@
 # Images list
-GET     /images          controllers.apiv1.APIv1Controller.images(query: Option[String] ?= None, offset: Int ?= 0, limit: Int ?= 25)
+GET     /images               controllers.apiv1.APIv1Controller.images(query: Option[String] ?= None, offset: Int ?= 0, limit: Int ?= 25)
 
 # A single image
-GET     /image/:id       controllers.apiv1.APIv1Controller.image(id: String)
+GET     /image/:id            controllers.apiv1.APIv1Controller.image(id: String)
+
+# Similiar images
```
Comment (Lucidiot) [inline conf/apiv1.routes]: All the code is gud but it's not `similiar` in English, but `similar`! (see [wiktionary](https://en.wiktionary.org/wiki/similar#English))

**Review by Lucidiot — CHANGES_REQUESTED** [review]
The code is mostly good except for one english mistake and one JS bad practice!

File: conf/apiv1.routes (hunk 3 above)
Comment (vaartis) [reply conf/apiv1.routes]: Oh. Yeah. In all other places it is correct, but not here.

File: assets/vue/ImageInfo.vue (hunk 1 above)
Comment (vaartis) [reply assets/vue/ImageInfo.vue]: Oh. Yeah i guess that's the purpose of using them.

File: conf/apiv1.routes (hunk 3 above)
Comment (Lucidiot) [reply conf/apiv1.routes]: Apparently not, just a Ctrl+F on this PR's changes shows 18 matches for `similiar`!

Comment (vaartis) [reply conf/apiv1.routes]: Oh. Wait. Not it's wrong everywhere..

**Review by Lucidiot — APPROVED** [review]
